File: java/src/tarehart/rlbot/steps/defense/GetOnDefenseStep.java
package tarehart.rlbot.steps.defense;

import mikera.vectorz.Vector3;
import tarehart.rlbot.AgentInput;
import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.SplineHandle;
import tarehart.rlbot.math.VectorUtil;
import tarehart.rlbot.planning.GoalUtil;
import tarehart.rlbot.planning.Plan;
import tarehart.rlbot.planning.SteerUtil;
import tarehart.rlbot.steps.Step;
import tarehart.rlbot.tuning.BotLog;

import java.util.Optional;

public class GetOnDefenseStep implements Step {
    private SplineHandle targetLocation = null;

    private Plan plan;

    public Optional<AgentOutput> getOutput(AgentInput input) {

        if (targetLocation == null) {
            init(input);
        }

        if (plan != null && !plan.isComplete()) {
            return plan.getOutput(input);
        }

        CarData car = input.getMyCarData();

        double distance = SteerUtil.getDistanceFromCar(car, targetLocation.getLocation());
        double secondsRemaining = distance / car.velocity.magnitude();

        if (!needDefense(input) || secondsRemaining < 1.5) {
            return Optional.empty();
        }

        Vector3 target;
        if (targetLocation.isWithinHandleRange(car.position)) {
            target = targetLocation.getLocation();
        } else {
            target = targetLocation.getFarthestHandle(input.ballPosition);
        }

        Optional<Plan> sensibleFlip = SteerUtil.getSensibleFlip(car, target);
        if (sensibleFlip.isPresent()) {
            BotLog.println("Front flip for defense", input.team);
            plan = sensibleFlip.get();
            plan.begin();
            return plan.getOutput(input);
        } else {
            return Optional.of(SteerUtil.steerTowardGroundPosition(car, target));
        }
    }

    private void init(AgentInput input) {
        targetLocation = GoalUtil.getOwnGoal(input.team).navigationSpline;
    }

    @Override
    public boolean isBlindlyComplete() {
        return false;
    }

    @Override
    public void begin() {
    }

    @Override
    public boolean canInterrupt() {
        return plan == null || plan.canInterrupt();
    }

    public static boolean needDefense(AgentInput input) {

        CarData car = input.getMyCarData();
        SplineHandle myGoal = GoalUtil.getOwnGoal(input.team).navigationSpline;

        boolean alreadyOnDefense = Math.abs(myGoal.getLocation().y - car.position.y) < 10;
        if (alreadyOnDefense) {
            return false;
        }

        Vector3 ballToGoal = (Vector3) myGoal.getLocation().subCopy(input.ballPosition);
        Vector3 ballVelocityTowardGoal = VectorUtil.project(input.ballVelocity, ballToGoal);

        double ballSpeedTowardGoal = ballVelocityTowardGoal.magnitude() * Math.signum(ballVelocityTowardGoal.dotProduct(ballToGoal));
        double wrongSidedness = getWrongSidedness(input);

        boolean needDefense = ballSpeedTowardGoal > 30 && wrongSidedness > 0 || ballSpeedTowardGoal > 10 && wrongSidedness > 10;
        return needDefense;

    }

    public static double getWrongSidedness(AgentInput input) {
        SplineHandle myGoal = GoalUtil.getOwnGoal(input.team).navigationSpline;
        double playerToBallY = input.ballPosition.y - input.getMyCarData().position.y;
        return playerToBallY * Math.signum(myGoal.getLocation().y);
    }

    @Override
    public String getSituation() {
        return "Getting on defense";
    }

}

File: java/src/tarehart/rlbot/steps/defense/GetOnDefenseStep.kt
package tarehart.rlbot.steps.defense

import tarehart.rlbot.AgentInput
import tarehart.rlbot.AgentOutput
import tarehart.rlbot.math.vector.Vector2
import tarehart.rlbot.planning.*
import tarehart.rlbot.routing.PositionFacing
import tarehart.rlbot.steps.NestedPlanStep
import tarehart.rlbot.steps.travel.ParkTheCarStep
import tarehart.rlbot.time.Duration
import tarehart.rlbot.time.GameTime

class GetOnDefenseStep @JvmOverloads constructor(private val lifespan: Double = DEFAULT_LIFESPAN // seconds
) : NestedPlanStep() {

    override fun getLocalSituation(): String {
        return "Getting on defense"
    }

    private var startTime: GameTime? = null

    override fun doComputationInLieuOfPlan(input: AgentInput): AgentOutput? {
        if (startTime == null) {
            startTime = input.time
        }

        val plan = Plan(Plan.Posture.DEFENSIVE).withStep(ParkTheCarStep { inp ->

            val goalPos = GoalUtil.getOwnGoal(inp.team).center
            val futureBallPosition = TacticsTelemetry[inp.playerIndex]?.futureBallMotion?.space ?:
                    inp.ballPosition

            val targetPosition = Vector2(
                    Math.signum(futureBallPosition.x) * CENTER_OFFSET,
                    goalPos.y - Math.signum(goalPos.y) * AWAY_FROM_GOAL)

            val targetFacing = Vector2(-Math.signum(targetPosition.x), 0.0)
            PositionFacing(targetPosition, targetFacing)
        })

        return startPlan(plan, input)
    }

    override fun shouldCancelPlanAndAbort(input: AgentInput): Boolean {
        val st = startTime ?: input.time
        return Duration.between(st, input.time).seconds > lifespan
    }

    companion object {
        private val CENTER_OFFSET = Goal.EXTENT * .5
        private val AWAY_FROM_GOAL = 3.0
        private val DEFAULT_LIFESPAN = 1.0
    }
}
